espnet2/train/checkpoint.py

```diff
@@ -0,0 +1,160 @@
+from typing import Any
+from typing import Iterable
+from typing import Optional
+from typing import Union
+
+import torch
+import torch.nn
+import torch.optim
+from typeguard import check_argument_types
+
+from espnet2.schedulers.abs_scheduler import AbsScheduler
+from espnet2.train.reporter import Reporter
+
+
+def resume(
+    model: torch.nn.Module,
+    optimizers: Iterable[torch.optim.Optimizer],
+    schedulers: Iterable[AbsScheduler],
+    reporter: Reporter,
+    output_dir: Union[str, Path],
+    resume_epoch: Optional[Union[int, str]],
+    resume_path: Optional[Union[str, Path]],
+    map_location: str,
+) -> None:
+    assert check_argument_types()
+    # For resuming: Specify either resume_epoch or resume_path.
+    #     - resume_epoch: Load from outdir/{}epoch/.
+    #     - resume_path: Load from the specified path.
+    # Find the latest epoch snapshot
+    if resume_epoch == "latest":
+        resume_epoch = 0
+        latest = None
+        for p in output_dir.glob("*epoch/timestamp"):
+            try:
+                n = int(p.parent.name.replace("epoch", ""))
+            except TypeError:
+                continue
+            with p.open("r") as f:
+                # Read the timestamp and comparing
+                date = f.read().strip()
+                try:
+                    date = datetime.strptime(date, "%Y-%m-%d %H:%M:%S")
+                except ValueError:
+                    continue
+            if latest is None or date > latest:
+                resume_epoch = n
+                latest = date
+
+    # If not found any snapshots, then nothing is done
+    if resume_epoch == 0:
+        resume_epoch = None
+
+    if resume_epoch is not None or resume_path is not None:
+        if resume_path is None:
+            resume_path = output_dir / f"{resume_epoch}epoch"
+            logging.info(f"--resume_epoch {resume_epoch}: Loading from {resume_path}")
+        else:
+            logging.info(f"--resume_path {resume_path}: Loading from {resume_path}")
+        load_checkpoint(
+            resume_path=resume_path,
+            model=model,
+            reporter=reporter,
+            optimizers=optimizers,
+            schedulers=schedulers,
+            map_location=map_location,
+        )
+
+
+def load_checkpoint(
+    resume_path: Union[str, Path],
+    model: torch.nn.Module,
+    reporter: Reporter,
+    optimizers: Iterable[torch.optim.Optimizer],
+    schedulers: Iterable[Optional[AbsScheduler]],
+    map_location: str,
+):
+    for key, obj in [("model", model), ("reporter", reporter)]:
+        _st = torch.load(resume_path / f"{key}.pth", map_location=map_location)
+        if obj is not None:
+            obj.load_state_dict(_st)
+    states = torch.load(resume_path / f"optim.pth")
+    for o, s, state in zip(optimizers, schedulers, states):
+        o.load_state_dict(state["optim"])
+        if s is not None:
+            s.load_state_dict(state["scheduler"])
+
+
+def save_checkpoint(
```

Comment:
Is there a way to control `save_checkpoint`?
If we save checkpoints every time, it may require a lot of storage... (espnet1 has such issue).

Reply:
I implemented `--keep_n_best_snapshots` for this (I'll change it to `--keep_n_best_checkpoints` to unify terms..)

- Except for the n-best snapshots, all files are removed at the every epochs.

https://github.com/espnet/espnet/blob/d21721fa1f9647d8c2a29c1c6ae8c5e2d2d765b4/espnet2/tasks/abs_task.py#L1055-L1074

- Regardless of this value, the snapshot of the latest epoch is always saved.

https://github.com/espnet/espnet/blob/d21721fa1f9647d8c2a29c1c6ae8c5e2d2d765b4/espnet2/tasks/abs_task.py#L1068

- The all checkpoints are kept among multiple criterions, e.g. `train.loss`, `eval.acc`.

https://github.com/espnet/espnet/blob/d21721fa1f9647d8c2a29c1c6ae8c5e2d2d765b4/espnet2/tasks/abs_task.py#L1058-L1062

- I'm not sure whether this is better, but it's also used as the number of  averaged model.

https://github.com/espnet/espnet/blob/d21721fa1f9647d8c2a29c1c6ae8c5e2d2d765b4/espnet2/tasks/abs_task.py#L835-L840

Another solution is something like `--save_checkpoints_interval`, but it is incompatible with averaging models, so I didn't append it.